include/revng/ABI/DependencyHelpers.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/ABI/Definition.h"
+#include "revng/Model/Binary.h"
+
+namespace abi {
+
+struct TypeDependencyManagementBase {
+public:
+  using TypeVector = llvm::SmallVector<UpcastablePointer<model::Type>, 16>;
+
+public:
+  const abi::Definition &ABI;
+
+protected:
+  TypeVector Dependencies;
+
+private:
+  model::Binary &Binary;
+
+private:
+  friend class ConverterBase;
+  TypeDependencyManagementBase(const abi::Definition &ABI,
+                               model::Binary &Binary) :
+    ABI(ABI), Binary(Binary) {}
+
+public:
+  TypeDependencyManagementBase(TypeDependencyManagementBase &&) = default;
+  virtual ~TypeDependencyManagementBase() { commit(); }
```

Comment:
> vague name and no docs;

Sorry.

> it's `virtual`, but not pure;

It's a destructor. Unless it's marked virtual, the children might not get deleted properly when they are stored in something like `std::unique_ptr<BaseClass>`.

> it's not used in any `unique_ptr`/pointer;

Oh, I see you're one step of me here. Luckily, I managed to avoid using polymorphism here.

> it has a `friend` class;

Just to allow the constructor being hidden - to avoid possible abuse. If you hate `friend`s that much, I don't mind making the constructor public and removing the `friend` declaration.

> inheritors do not add any new fields;

Yep, I luckily managed to get rid of any special fields - making the usage a lot simpler. At this point, maybe even inheritance is not necessary and it can just be a property of its "children", but I feel like it's easier to work with it this way.

> Are you sure inheritance and being `virtual` is the right tool here?

So, to summarize, no, inheritance is not necessary, BUT I feel like it's easier to reason about children with how similar they are in terms of the shared parents. Consider that they are going to be placed in two different projects after all - the more explicitly we can show that "this stuff is shared", the easier the connection is going to be to maintain.

> `virtual`

Of course the virtual destructor is not required here since inheritors don't override it anyway, BUT I feel like that's a useful bit of future-proofing: this way a suddenly added member to one of them has no chance of causing any trouble. Sure we're paying a bit of runtime cost each time an object is deleted, but I can't imagine it causing a bottleneck here: there are not many of these objects anyway.

> What are you trying to achieve here?

A clear and decoupled way to manage types that are "temporarily out of model as we might want to discard them".

> Can you merge this class with `Converter`?

Actually it was merged in the last time you reviewed the `Converter`, for details see [this thread](https://github.com/revng/revng/pull/232#discussion_r998325888).

P. S. Gosh, I think this is my new record for the longest response 😅.

Reply:
We can make it `virtual` if needed.
If you need inheritance to share implementation details I'm OK with that, but the `...Base` suffix + `virtual`ity is misleading.

But think again: inheritors do not add new member and they don't seem to manipulate directly the set of "temporary types" (which is a nice solution btw!), so probably the simplest thing is just to go for composition.
Consider moving the methods of the inherited classes into `Converter`s and turn into a single-purpose class for temporary types that need to be committed.

Something like:

```c++
struct Converter {
  TypeDependencyManagementBase TemporaryTypes;

  void tryConvertingRegisterArguments() {
    TemporaryTypes.doSomething();
  }
};

Reply:
I'll see what I can do, but I feel like the temporary-type-adjacent methods are not trivial to just  move back, since there's more than a single place where they get merged back (because of the fact that there's more than one entry point to one of the converters - `Layout` needs a wider interface then just `toRaw`.

Reply:
Can't you expose `Converter` to `Layout` too?

Reply:
That's not what I mean. `Layout` uses "internal" methods of the `Converter`. As the result, the converter is forced to have more than one `TypeBucket`.

Reply:
I need to look into this again.

Reply:
Gotcha, I won't push anything related to it in the meantime then. Please let me know when you're done.

Reply:
I've reviewed this and made [a graph](https://pad.rev.ng/mGcRiN5-RXuQULzVijv6Hg#).

Some observations:
* `ConverterBase` does practically nothing. It just returns an instance of `TypeDependencyManagementBase`, which is very confusing because usually `...Base` classes are abstract classes and you pass around a pointer to them to use virtual methods. But:
  * This is not the case for `ConverterBase` (can't find any usage as a pointer).
  * This is not the case for `TypeDependencyManagementBase` either
* The two `TypeDependencyManagement` are constructed from their base class. Which is very weird.

I suggest to:
1. Drop `ConverterBase`
2. Drop both `revng::TypeDependencyManager` and `revng::c::TypeDependencyManager`, their methods can go in `revng::Converter` and `revng::c::Converter`
3. Add a field of type `TypeDependencyManagementBase` to `Converter`s.
4. Rename the `Converters` to `ToRawConverter` and `ToCABIConverter` or `FunctionTypeLowerer` and `FunctionTypeLifter`.
5. Rename `TypeDependencyManagementBase` to `TypeBucket`/`TypeBin`.

This is more or less what I already suggested. 
You said:
> there's more than one entry point to one of the converters - Layout needs a wider interface then just toRaw

I don't see what would be the problem in leaving those methods in `ToRawConverter`/`FunctionTypeLowerer`.

> That's not what I mean. Layout uses "internal" methods of the Converter. As the result, the converter is forced to have more than one TypeBucket.

Or you could just have two instances of the `Converter`, depending on what you need. Would that be an issue?

Reply:
Ok, I'm convinced. Thanks for looking so deeply into this.

Reply:
Updates?

Reply:
Please take a look at the updated version: https://github.com/revng/revng/pull/232/commits/7c6febcdecc7aed7c05db2fc372bb9f00199f283

Reply:
Since the new version of this thing seems to have passed the review, I'm closing this thread.